### ISSUE TITLE
Different query values on 2.2.* comparing with 2.1.*

### ISSUE THREAD

**Issue by ijwmiranda** [orientechnologies/orientdb#7394]
## Actual behavior  
After upgrading to 2.2.19 from 2.1.* 

The '<>' operator don't work as we expected on the new version.

## Steps to reproduce  

The following queries returns differents results:

```
select count(*) from (TRAVERSE out from E) -> Number of entries 81656

select count(*) from (TRAVERSE out from E) where in.@rid <> out.@rid -> Number of entries 73453

select count(*) from (TRAVERSE out from E) where not(in.@rid = out.@rid) -> Number of entries 81394

select count(*) from (TRAVERSE out from E) where not(in = out) -> Number of entries 81394

select count(*) from (TRAVERSE out from E) where in <> out -> Number of entries 51835

select count(*) from (TRAVERSE out from E) where in = out -> Number of entries 262

select count(*) from (TRAVERSE out from E) where in.@rid = out.@rid -> Number of entries 262
```

Apparentely on the new version, the '<>' or '!=' don't evaluate the equality between two entries, can we have a rational that explains the above results?

Thanks in advance

**Comment by luigidellaquila** [comment]
Hi @ijwmiranda 

I tried to reproduce the problem, but with no luck.
My script is:

```
create vertex v set name = 'a'
create vertex v set name = 'b'
create edge E from (select from v where name = 'a') to (select from V)

select count(*) from (traverse out from E) where out.@rid <> in.@rid
```

and it returns correct results

Do you have a dataset to reproduce it?

Thanks

Luigi

**Comment by ijwmiranda** [comment]
Hi Luigi,

Thanks for the quick answer, we just test with the provided database: GratefulDeadConcerts.

I'm not sure if the database changed but i think not, we use the following versions:

2.1.24
2.2.19

The problem is self explained on the following queries:

Query1: `select count(*) from (TRAVERSE out from E)`
2.1.24 Entries Number: 8633
2.2.19 Entries Number: 8633
Note: Total number of entries

Query2: `select count(*) from (TRAVERSE out from E) where in.@rid <> out.@rid`
2.1.24 Entries Number: 8049
2.2.19 Entries Number: 8049

Query3: `select count(*) from (TRAVERSE out from E) where in <> out`
2.1.24 Entries Number: 8049
2.2.19 Entries Number: **7097**

Query4: `select count(*) from (TRAVERSE out from E) where ( in.@rid <> out.@rid and not(in <> out) )`
2.1.24 Entries Number: 0
2.2.19 Entries Number: **952**

Thanks

**Comment by luigidellaquila** [comment]
Great, thank you for the info, I'm checking it

Luigi

**Comment by luigidellaquila** [comment]
Ok, I could reproduce the problem.
The bug seems to be in the binary representation of a record and how it's managed by the binary comparator: when it tries to do a `ORecordSerializerBinaryV0.readOptimizedLink()` it throws an `ArrayIndexOutOfBoundsException`

@tglman it's for you. Easy steps to reproduce and find the problem:
- 2.2.x snapshot, GratefulDeadConcerts db
- add a breakpoint on ArrayIndexOutOfBoundsException
- run `select from (TRAVERSE out from E) where in <> out`

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Ok, never mind, the problem was actually in the binary comparator, I'm fixing it

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi @ijwmiranda 

I pushed a fix for this on branch 2.2.x and develop, the fix will be released with next 2.2.20

Thanks

Luigi
